### PR TITLE
Add a DPP intro in amdgpu_kernel_optimization_guide.md

### PR DESCRIPTION
**Pull request by knwng**

### CODE REVIEW

File: docs/amdgpu_kernel_optimization_guide.md [hunk 1]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+Since DPP is an instruction modifier instead of an instruction itself, there are
+tremendous number of combinations of VALU instructions and DPP. To solve that, `rocdl.update.dpp`
+and `amdgpu.dpp` are designed to be a wrapper of `v_mov_b32_dpp` instruction. And it depends
+on LLVM compiler to fuse it with the subsequent VALU instruction **with best efforts**.
+
+For example, `v_mov_b32_dpp` + `v_add_f32_e32` might be fused into `v_add_f32_dpp`.
+
+There are plenty of constraints stopping an instruction from being merged.
+For example, if either the `bank_mask` or the `row_mask` is not `0xf`, it can't be fused.
+You can check the [GCNDPPCombine::combineDPPMov](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.3/llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp#L522) function to see how it works.
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:375]: Use permalinks for GitHub links that point to specific files / line numbers

File: docs/amdgpu_kernel_optimization_guide.md [hunk 2]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+
+### Comparison
+
+To summarize, there's no free lunch. You can't have both performance and generality
+at the same time.
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:380]:
```suggestion
To summarize, there's no free lunch: instruction's expressivity comes at the expense of performance.
```

File: docs/amdgpu_kernel_optimization_guide.md [hunk 3]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+
+The performance ranking is like:
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:382]:
```suggestion
The relative performance of cross-lane instructions is as follows:
```

File: docs/amdgpu_kernel_optimization_guide.md [hunk 4]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+
+DPP > `ds_swizzle` >= `ds_permute`/`ds_bpermute`
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:384]:
```suggestion
DPP > `ds_swizzle` >= `ds_permute` > `ds_bpermute`
```

Comment (kuhar) [reply docs/amdgpu_kernel_optimization_guide.md:384]: permute should be faster than bpermute

File: docs/amdgpu_kernel_optimization_guide.md [hunk 5]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+
+while the generality ranking is the opposite:
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:386]:
```suggestion
while the generality ranking is the reverse:
```

File: docs/amdgpu_kernel_optimization_guide.md [hunk 6]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+
+DPP < `ds_swizzle` < `ds_permute`/`ds_bpermute`
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:388]:
```suggestion
DPP < `ds_swizzle` < ds_bpermute` < `ds_permute`
```
bpermute is easier to use than permute

File: docs/amdgpu_kernel_optimization_guide.md [hunk 7]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+
+This table summarized some useful facts. The latency data is collected from an experiment
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:390]:
```suggestion
This table presents the approximate instruction latency, collected experimentally
```

File: docs/amdgpu_kernel_optimization_guide.md [hunk 8]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+on Fused Softmax kernel with [rocprofv2](https://github.com/ROCm/rocprofiler?tab=readme-ov-file#plugin-support):
+
+| Instructions           | MLIR Op                      | Hardware     | latency/#cycles |
+| ---------------------- | ---------------------------- | ------------ | --------------- |
+| ds_permute/ds_bpermute | rocdl.ds_bpermute            | LDS hardware | ~50*            |
+| ds_swizzle             | rocdl.ds_swizzle             | LDS hardware | ~50*            |
+| DPP                    | rocdl.update.dpp, amdgpu.dpp | VALU         | 4~12            |
+
+*: For ds_permute/ds_bpermute and ds_swizzle, the latency includes the instruction itself
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:399]:
```suggestion
For `ds_permute`/`ds_bpermute` and `ds_swizzle`, the latency includes the instruction itself
```

File: docs/amdgpu_kernel_optimization_guide.md [hunk 9]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+The performance ranking is like:
+
+DPP > `ds_swizzle` >= `ds_permute`/`ds_bpermute`
+
+while the generality ranking is the opposite:
+
+DPP < `ds_swizzle` < `ds_permute`/`ds_bpermute`
+
+This table summarized some useful facts. The latency data is collected from an experiment
+on Fused Softmax kernel with [rocprofv2](https://github.com/ROCm/rocprofiler?tab=readme-ov-file#plugin-support):
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:391]: Also mention the hardware. I assume this is on mi300?

**Review by kuhar — COMMENTED** [review]
Very cool, thanks for writing this up @knwng!

Also please update the header at the very top of the file.

File: docs/amdgpu_kernel_optimization_guide.md (hunk 9 above)
Comment (knwng) [reply docs/amdgpu_kernel_optimization_guide.md:391]: Yes

File: docs/amdgpu_kernel_optimization_guide.md [hunk 10]
```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
...
+
+For details, please check [MI300 ISA Reference Guide](https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-mi300-cdna3-instruction-set-architecture.pdf) and
+[RDNA3 ISA Reference Guide](https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf)
+
+### How to use them in MLIR
+
+Each instruction has its corresponding Op in MLIR(except `ds_permute`):
+- `ds_bpermute`: `rocdl.ds_bpermute`
+- `ds_swizzle`: `rocdl.ds_swizzle`
+- DPP: `rocdl.update.dpp`, `amdgpu.dpp`(a slim wrapper of `rocdl.update.dpp` with more comprehensive user interface)
```
[the remark this answers is not in the document]
Comment (knwng) [reply docs/amdgpu_kernel_optimization_guide.md:362]: The main difference is `amdgpu.dpp` use some enums to replace the magic numbers representing DppCtrl.

[the remark this answers is not in the document]
Comment (knwng) [reply docs/amdgpu_kernel_optimization_guide.md:362]: Added something to explain this.

File: docs/amdgpu_kernel_optimization_guide.md [hunk 11]
```diff
@@ -293,3 +293,107 @@ forms a *clause* that translates to a single data fabric transaction.
...
+The relative performance of cross-lane instructions is as follows:
+
+DPP > `ds_swizzle` >= `ds_permute` > `ds_bpermute`
+
+while the generality ranking is the reverse:
+
+DPP < `ds_swizzle` < `ds_permute` < `ds_bpermute`
+
+This table presents the approximate instruction latency, collected experimentally
+on Fused Softmax kernel with [rocprofv2](https://github.com/ROCm/rocprofiler?tab=readme-ov-file#plugin-support) on MI300 GPU:
```
Comment (kuhar) [inline docs/amdgpu_kernel_optimization_guide.md:390]:
```suggestion
on Fused Softmax kernel with [rocprofv2](https://github.com/ROCm/rocprofiler?tab=readme-ov-file#plugin-support) on the MI300 GPU:
```

**Review by kuhar — APPROVED** [review]
Looks good. Could you rewrap lines before we merge this? In vscode there's a command for this under Alt + Q.

**Pull request merged**